### PR TITLE
added sunbird api for 2.10

### PR DESCRIPTION
**Pull request by aman-tunerlabs**

### CODE REVIEW

File: .env.sample [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+SLACK_COMMUNICATIONS_ON_OFF = "OFF"
+SLACK_EXCEPTION_LOG_URL = "https://hooks.slack.com/services/*******/*********/**************"
+SLACK_TOKEN = "xoxp-XXXXXXXXXX-XXXXXXXXX-XXXXXXXXX-XXXXXXXXX"
+
+URL_PREFIX = "api/v1"
+HEALTH_CHECK_URL = "/ping"
+
+LOGGER_DIRECTORY = "logs"
+
+# APi Doc
```
Comment (aks30) [inline .env.sample]: @aman-tunerlabs - Rename it to API Doc

File: .env.sample [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+BODY_PARSER_LIMIT = "50MB"
+
+REPORTS_PATH = "/public/reports"
+
+# CLOUD SERVICES 
+
+CLOUD_STORAGE = "GC"
+GCP_BUCKET_NAME = "sl-dev-storage"
+QR_CODE_BUCKET_NAME = "sl-dev-storage"
```
Comment (aks30) [inline .env.sample]: @aman-tunerlabs -Missing GCP config file

File: module/bodh/request/validator/v1.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * created-date : 06-04-2020
+ * Description : Bodh request validator.
+ */
+
+module.exports = (req) => {
+
+    let bodhRequestValidator = {
+        
+        middleware : function () {
+            req.checkBody('url').exists().withMessage("required url");
```
Comment (aks30) [inline module/bodh/request/validator/v1.js:13]: @rakeshSgr - Check if the URL has a valid sunbird endpoint. Refer to Kendra for utils function

File: module/bodh/search/validator/v1.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+module.exports = (req) => {
+
+    let searchRequestValidator = {
+        
+        middleware: function () {
+            req.checkBody('url').exists()
+            .notEmpty()
+            .custom((value) => (value) ? gen.utils.checkIfURLIsSunbirdAPI(value) : false)
```
Comment (aks30) [inline module/bodh/search/validator/v1.js:8]: @rakeshSgr - Utils constant will change.

File: controllers/v1/learningResources.js [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+  static get name() {
+    return "learningResources";
+  }
+
+  /**
+  * @api {get} /sunbird/api/v1/learningResources/list?limit=10&page=1 
+  * To get list of learning resources
+  * @apiVersion 1.0.0
+  * @apiGroup Learning Resources
+  * @apiHeader {String} X-authenticated-user-token Authenticity token
```
Comment (aks30) [inline controllers/v1/learningResources.js:25]: @rakeshSgr - Let this be available only if the internal access token is provided.

File: controllers/v1/token.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+  static get name() {
+    return "token";
+  }
+
+  /**
+  * @api {get} /sunbird/api/v1/token/verify
+  * To verify the access token
+  * @apiVersion 1.0.0
+  * @apiGroup Token
+  * @apiHeader {String} X-authenticated-user-token Authenticity token
```
Comment (aks30) [inline controllers/v1/token.js]: @rakeshSgr - Only for internal-access-token giving requests.

File: controllers/v1/token.js [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+  * @apiSampleRequest /sunbird/api/v1/token/verify
+  * @apiUse successBody
+  * @apiUse errorBody
+  * @apiParamExample {json} Response:
```
Comment (aks30) [inline controllers/v1/token.js:34]: @rakeshSgr - Add request body as well

File: generics/constants/api-responses.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  "CONTENT_FILE_REQUIRED": "Required content file name",
+  "CONTENT_UPLOADED_SUCCESSFULLY": "Successfully uploaded content",
+  "COULD_NOT_UPLOAD_CONTENT": "Could not upload content",
+  "COULD_NOT_UNZIP_FOLDER": "Unable to unzip folder",
+  "COULD_NOT_RENAME_FILE": "Could not rename a file",
+  "COULD_NOT_ZIP_FOLDER": "Unable to zip folder",
+  "COULD_NOT_SAVE_ZIP_FILE": "Could not save zip file",
+  "QR_CODE_DATA_SIZE": "",
+  "KENDRA_SERVICE_DOWN": "Kendra service is down",
+
```
Comment (aks30) [inline generics/constants/api-responses.js]: @rakeshSgr - Remove extra lines as double quotes from the key

File: generics/constants/common.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    AWS_SERVICE: "AWS",
+    GOOGLE_CLOUD_SERVICE: "GC",
+    AZURE_SERVICE: "AZURE",
+    "FAILED" : "Failed.",
+    "SUCCESS" : "Success.",
+    "OK" : "OK",
+    "POST_METHOD" : "POST",
+    "BODH_CONTENT_TYPE" : "Resource",
+    "BODH_MIME_TYPE" : "application/vnd.ekstep.html-archive",
+    "FORM_DATA_CONTENT_TYPE" : "application/x-www-form-urlencoded",
```
Comment (aks30) [inline generics/constants/common.js]: @rakeshSgr -Maintain common standard of no double quotes in keys

File: generics/constants/endpoints.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    SUNBIRD_GENERATE_DIALCODE : "/api/dialcode/v1/generate",
+    SUNBIRD_PUBLISH_DIALCODE : "/api/dialcode/v1/publish",
+    SUNBIRD_DIALCODE_STATUS : "/api/dialcode/v1/read",
+    SUNBIRD_CONTENT_LINK : "/api/dialcode/v1/content/link",
+    SUNBIRD_PUBLISH_CONTENT : "/api/content/v1/publish",
+    SUNBIRD_USER_READ : "/api/user/v1/read",
+    SUNBIRD_INDEX_SYNC : "/api/data/v1/index/sync",
+    SUNBIRD_CREATE_CONTENT : "/api/content/v1/create",
+    SUNBIRD_UPLOAD_CONTENT : "/api/content/v1/upload",
+
```
Comment (aks30) [inline generics/constants/endpoints.js]: @rakeshSgr - Remove extra lines everywhere

File: generics/helpers/slack-communications.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+              value: errorMessage[objValue],
+              short: false
+            })
+          }
+        })
+
+
+        fieldsData.push({
+          title: "Environment",
+          value: process.env.NODE_ENV,
```
Comment (aks30) [inline generics/helpers/slack-communications.js]: @rakeshSgr - This env value won't be there

File: module/files/helper.js [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+   */
+
+  static saveZipFile(name, data) {
+    return new Promise(async (resolve, reject) => {
+      try {
+        if (!fs.existsSync(`${PROJECT_ROOT_DIRECTORY}${process.env.ZIP_PATH}`)) {
+          fs.mkdirSync(`${PROJECT_ROOT_DIRECTORY}${process.env.ZIP_PATH}`)
+        }
+
+        let zipFileName = `${PROJECT_ROOT_DIRECTORY}${process.env.ZIP_PATH}/${name}`
```
Comment (aks30) [inline module/files/helper.js:219]: @rakeshSgr - I think ZIP_PATH is not set in env

File: package.json [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    "csvtojson": "^2.0.10",
+    "dotenv": "^8.2.0",
+    "ejs": "^3.1.3",
+    "express": "^4.17.1",
+    "express-cassandra": "^2.3.2",
+    "express-fileupload": "^1.1.6",
+    "express-validator": "^5.3.1",
+    "fn-args": "^4.0.0",
+    "form-urlencoded": "^4.1.3",
+    "fs": "0.0.1-security",
```
Comment (aks30) [inline package.json:49]: @rakeshSgr - I think one fs-extra module is used in migrations, which is not included here

File: module/token/helper.js [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+* @method
+* @name  verify
+* @param {String} token - user access token.
+* @returns {json} Response consists of user token details
+*/
+static verify(token) {
+    return new Promise(async (resolve, reject) => {
+       try {
+      
+       apiInterceptor.validateToken(token, async function (err, tokenData) {
```
Comment (aks30) [inline module/token/helper.js]: @rakeshSgr - Let the API of verification token response from the controller be like 
`{
    "message": "Token verified",
    "status": 200,
    "result": {
        "isValid": true/false,
         "userInformation" : {
 // If isValid token return the jwt info
          }
}
`

File: module/token/helper.js [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * name : token/helper.js
+ * author : Rakesh Kumar
+ * created-date : 21-july-2020
+ * Description : Related to user token validation
+ * to cloud service.
+ */
+
+// Dependencies
+const apiInterceptor = require(PROJECT_ROOT_DIRECTORY + "/generics/middleware/apiInterceptor");
```
Comment (aks30) [inline module/token/helper.js]: @rakeshSgr - Use constants for generic helpers and generic middlewares

File: module/token/helper.js [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+            if (userDetails.responseCode == "OK") {
+                resolve({ message:CONSTANTS.apiResponses.VALID_TOKEN, result:userDetails.result.response })
+            }else{
+                return resolve({ status : HTTP_STATUS_CODE["unauthorized"].status , message:CONSTANTS.apiResponses.INVALID_TOKEN });
+            }
+        }
+        });
+    
+       } catch (error) {
+            return reject(error);
```
Comment (aks30) [inline module/token/helper.js]: @rakeshSgr - Let all helpers follow the below structure

In try 

```
{
    success : true,
    message : "Success message text",
    data : true/object/array
}
```

In Catch 

```
{
    success : false,
    message : "Error message text explaining the error",
    data : false
}
```

File: routes/index.js [hunk 17]
```diff
@@ -0,0 +1,173 @@
+/**
+ * name : index.js
+ * author : Aman Karki
+ * Date : 20-July-2020
+ * Description : All routes.
+ */
+
+// Dependencies
+const authenticator = require(PROJECT_ROOT_DIRECTORY + "/generics/middleware/authenticator");
```
Comment (aks30) [inline routes/index.js]: @rakeshSgr - Please use constants for middlewares and helpers

File: module/learningResources/helper.js [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+    * @param {String} token - user access token.
+    * @returns {json} Response consists of list of filters
+    */
+    static filtersList(token) {
+        return new Promise(async (resolve, reject) => {
+            try {
+                let category = await sunbirdService.filtersList(token);
+                if (category) {
+                    if (category) {
+                        resolve({ message: CONSTANTS.apiResponses.FILTERS_FOUND, result: category })
```
Comment (aks30) [inline module/learningResources/helper.js]: @rakeshSgr - Can we store filters in session?

File: module/learningResources/helper.js [hunk 19]
```diff
@@ -0,0 +1,82 @@
+
+
+let sunbirdService =
```
Comment (aks30) [inline module/learningResources/helper.js]: @rakeshSgr - Add names and other meta. Also, use constant

File: generics/services/sunbird.js [hunk 20]
```diff
@@ -0,0 +1,598 @@
...
+            "filters": {
+                "contentType": ["Resource"],
+            },
+            "limit": limit,
+            "mode": "soft",
+            "offset": offset -1
+          }
+
+          if(filters){
+              if(filters["board"]){
```
Comment (aks30) [inline generics/services/sunbird.js]: @rakeshSgr - Would suggest making it dynamic, what the keys are passed in filter object,loop through all and add a filter in request body.

File: generics/services/sunbird.js [hunk 21]
```diff
@@ -0,0 +1,598 @@
...
+                "authorization": process.env.AUTHORIZATION,
+                "x-authenticated-user-token": token
+            }
+        };
+
+        if (requestType != "GET") {
+            options['json'] = { request: requestBody };
+        }
+
+        url =process.env.SUNBIRD_BASE_URL + url;
```
Comment (aks30) [inline generics/services/sunbird.js]: @rakeshSgr - Consider using request module as in above functions

File: generics/helpers/slack-communications.js (hunk 11 above)
Comment (rakeshSgr) [reply generics/helpers/slack-communications.js]: It has been used in other places as well. Shall I remove it or add it to .env?

File: generics/services/sunbird.js (hunk 21 above)
Comment (rakeshSgr) [reply generics/services/sunbird.js]: Need clarification

File: controllers/v1/qr-codes.js [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+    static get name() {
+        return "qr-codes";
+    }
+
+
+  /**
+     * @api {get} /sunbird/api/v1/qr-code/image/:uniquecode 
+     * Get the image link of the qr code
+     * @apiVersion 1.0.0
+     * @apiGroup qrCodes
```
Comment (aks30) [inline controllers/v1/qr-codes.js]: @rakeshSgr - API group has to start with a capital letter.

File: controllers/v1/token.js [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+  * To verify the access token
+  * @apiVersion 1.0.0
+  * @apiGroup Token
+  * @apiBody {String} X-authenticated-user-token Authenticity token
+  * @apiSampleRequest /sunbird/api/v1/token/verify
+  * @apiUse successBody
+  * @apiUse errorBody
+  * @apiParamExample {json} Request:
+  * {
+  *   "token": "eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICItRjZVRW5nVHdlZFg1UlZaSkNnRFZWZ1IxYURjYmFrUUplams0cW
```
Comment (aks30) [inline controllers/v1/token.js]: @rakeshSgr - Please remove the token, give a sample with random characters.

File: controllers/v1/token.js [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+   * @class
+*/
+module.exports = class Token {
+
+  static get name() {
+    return "token";
+  }
+
+  /**
+  * @api {get} /sunbird/api/v1/token/verify
```
Comment (aks30) [inline controllers/v1/token.js]: @rakeshSgr - Change it to post call

File: controllers/v1/token.js [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+  * To verify the access token
+  * @apiVersion 1.0.0
+  * @apiGroup Token
+  * @apiBody {String} X-authenticated-user-token Authenticity token
```
Comment (aks30) [inline controllers/v1/token.js]: @rakeshSgr - Change it key as token

File: generics/helpers/slack-communications.js [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+  * @name sendMessageToSlack
+  * @param {Object} errorMessage - error message to be sent to slack.
+  * @returns {Promise} returns a promise.
+*/
+
+let sendMessageToSlack = function (errorMessage) {
+  return new Promise(async (resolve, reject) => {
+    try {
+
+      if (SLACK_COMMUNICATION_ON_OFF === "ON" &&
```
Comment (aks30) [inline generics/helpers/slack-communications.js]: @rakeshSgr - Move this check in envVariables

File: generics/helpers/slack-communications.js [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+ */
+
+//dependencies
+
+const request = require('./http-request');
+const SLACK_COMMUNICATION_ON_OFF = process.env.SLACK_COMMUNICATIONS_ON_OFF;
+const SLACK_TOKEN = process.env.SLACK_TOKEN
+const SLACK_HOOK = process.env.SLACK_EXCEPTION_LOG_URL;
+const ERROR_MESSAGES_TO_SLACK = 
+(!process.env.ERROR_MESSAGE_TO_SLACK ||
```
Comment (aks30) [inline generics/helpers/slack-communications.js]: @rakeshSgr - Add default values in envVariables

File: generics/middleware/authenticator.js [hunk 28]
```diff
@@ -0,0 +1,180 @@
...
+
+  removedHeaders.forEach(function (e) {
+    delete req.headers[e];
+  });
+
+  var token = req.headers["x-authenticated-user-token"];
+  if (!req.rspObj) req.rspObj = {};
+  var rspObj = req.rspObj;
+
+  if (req.path.includes("slack")) {
```
Comment (aks30) [inline generics/middleware/authenticator.js]: @rakeshSgr - No need of this code.

File: generics/middleware/authenticator.js [hunk 29]
```diff
@@ -0,0 +1,180 @@
...
+  let guestAccessPaths = ["bodh/search","bodh/request"];
+  await Promise.all(guestAccessPaths.map(async function (path) {
+    if (req.path.includes(path)) {
+      next();
+      return;
+    }
+  }));
+
+  let internalAccessApiPaths = ["token/verify","keywords"];
+  let performInternalAccessTokenCheck = false;
```
Comment (aks30) [inline generics/middleware/authenticator.js:94]: @rakeshSgr - Handle case of internal access token and x-authenticated user token both are required.

File: module/learningResources/helper.js [hunk 30]
```diff
@@ -0,0 +1,90 @@
...
+                    board: board,
+                    gradeLevel: gradeLevel,
+                    subject: subject,
+                    medium: medium
+                }
+                sss
+                let learningResources = await sunbirdService.learningResources(token, pageSize, pageNo, filters, sortBy);
+                if (learningResources) {
+
+                    if (learningResources.content) {
```
Comment (aks30) [inline module/learningResources/helper.js]: @rakeshSgr - Modify it as per the helper function return structure we defined.

File: module/learningResources/helper.js [hunk 31]
```diff
@@ -0,0 +1,90 @@
...
+    * To get filters list
+    * @method
+    * @name  filtersList
+    * @param {String} token - user access token.
+    * @returns {json} Response consists of list of filters
+    */
+    static filtersList(token) {
+        return new Promise(async (resolve, reject) => {
+            try {
+                let filters = sessions.get("Filters");
```
Comment (aks30) [inline module/learningResources/helper.js]: @rakeshSgr - Filter key has to meaningful and unique to the set of filters.

File: module/learningResources/helper.js [hunk 32]
```diff
@@ -0,0 +1,90 @@
...
+                if (!filters) {
+                    filters = await sunbirdService.filtersList(token);
```
Comment (aks30) [inline module/learningResources/helper.js:78]: @rakeshSgr - Set filters in session here

File: module/token/helper.js [hunk 33]
```diff
@@ -0,0 +1,72 @@
...
+                        if (userDetails.responseCode == "OK") {
+                            resolve({
+                                message: CONSTANTS.apiResponses.VALID_TOKEN,
+                                result: {
+                                    isValid: true,
+                                    userInformation: userDetails.result.response
+                                }
+                            })
+                        } else {
+                            return resolve({
```
Comment (aks30) [inline module/token/helper.js:55]: @rakeshSgr - Follow the same structure as discussed.

File: generics/services/sunbird.js [hunk 34]
```diff
@@ -0,0 +1,590 @@
+/**
+ * name : bodh.js
```
Comment (aks30) [inline generics/services/sunbird.js]: @rakeshSgr - File name is wrong.

File: generics/services/sunbird.js [hunk 35]
```diff
@@ -0,0 +1,590 @@
...
+ * author : Aman Jung Karki
+ * Date : 11-Nov-2019
+ * Description : All bodh service related information.
+ */
+
+//dependencies
+
+const request = require('request');
+const shikshalokamService = require(PROJECT_ROOT_DIRECTORY + "/generics/helpers/shikshalokam");
```
Comment (aks30) [inline generics/services/sunbird.js]: @rakeshSgr - Use constant of generic helpers.

File: generics/services/sunbird.js [hunk 36]
```diff
@@ -0,0 +1,590 @@
...
+    return new Promise(async (resolve, reject) => {
+
+        try {
+
+            const learningResourceUrl = CONSTANTS.endpoints.GET_RESOURCES;
+
+            let requestBody = {
+                "source": "web",
+                "name": "Resource",
+                "facets": ["board", "gradeLevel", "subject", "medium"],
```
Comment (aks30) [inline generics/services/sunbird.js]: @rakeshSgr - Make this dynamic

File: generics/services/sunbird.js [hunk 37]
```diff
@@ -0,0 +1,590 @@
...
+
+        function callback(err, data) {
+            if (err) {
+                return reject({
+                    message: CONSTANTS.apiResponses.SUNBIRD_SERVICE_DOWN
+                });
+            } else {
+
+
+                if (data.statusCode == HTTP_STATUS_CODE.ok.status) {
```
Comment (aks30) [inline generics/services/sunbird.js:475]: @rakeshSgr - Plan for handling the scenario where if sunbird doesn't respond within 3 sec, give an error if it does. Make the timeout value configurable.

**Pull request merged**